[MRG] add vol to surf by sampling between nested surfaces

jeromedockes opened the pull request:
This PR introduces a new method for volume to surface projections: the user
provides two nested surfaces with one-to-one correspondence between nodes, and
values are sampled between each pair of nodes. IIRC @bthirion asked for this a
couple of times. Still WIP but I would appreciate feedback on the suggested
interface (add an `inner_mesh` optional parameter), and to know where I can
download white matter surface meshes for the fsaverage5 and fsaverage surfaces
to add fetchers for these.

~~Also, to be able to test surface-related functions on more realistic data, this
PR adds a couple of utils in `data_gen`: one to generate random low-resolution
images in mni space, and one to extract a (bad) mesh from the MNI mask shipped
with nilearn at a chosen resolution. This enables getting low-resolution
(possibly nested) meshes for testing without needing to add more data to the
package or to require network access, but while still having something that
resembles more a brain surface in size, position and topology than random points
and random triangles. (a few examples below)~~

![example_plots](https://user-images.githubusercontent.com/9196501/83808755-54b07b00-a683-11ea-9034-cb1f347bfd07.png)

bthirion commented:
My first thought is not to go into the business of mesh generation. I guess that fsaverage is used 90% of the time and for other cases, people will reuse the freesurfer generated stuff.

bthirion commented on nilearn/surface/surface.py:
```diff
@@ -442,6 +475,16 @@ def vol_to_surf(img, surf_mesh,
+        across the gray matter thickness from node i in `inner_mesh`. Image
+        values for index i are then sampled along the line joining these two
+        points.
+
```
My understanding is that quite often, people want to select the depth at which they're sampling (layered fMRI). This only makes sense for high resolution data of course. 
So I would define the sampling scheme along the cortical depths by a list of values between 0 and 1: [0]=white, [1]=pial, [.5]=middle surface, [0, 0.1,..,1]= average across 10 intermediate surfaces.

jeromedockes replied:
ok! should that be an extra parameter then? or we can deprecate the "ball" sampling strategy and apply this depth selection either to the vertex normal or to the between-mesh segment, depending on whether a second surface is provided or not. In the single mesh /fixed depth case users may also want to select the offsets of sample points. for now they can change the scale but the points are always centered on the vertex

bthirion replied:
I think that the ball think is a very good idea if you have only one surface. Ideally, I would like to make it possible to use both heuristics. It is also better for backward compatibility.
The problem is how to have both in the same API, and making it clear !

jeromedockes commented:
> My first thought is not to go into the business of mesh generation. I guess that fsaverage is used 90% of the time and for other cases, people will reuse the freesurfer generated stuff.

note the mesh generation is intended for internal unit tests only, I'm
not suggesting to expose it to users. We can use fsaverage5 for all
tests as it is shipped with nilearn and hence doesn't need downloading,
but it might be a bit large and slow down some tests unnecessarily, and
it may not be ideal to always test on one single mesh. also that would
require also including the fsaverage white matter with the code -- but
we may want to do that anyway.

bthirion commented:
Also, there is a glitch that remains to be fixed. I could not find the issue, but it was something like a half voxel shift that needs to be addressed for freesurfer meshes.  IIRC this resulted in LR inaccuracies typically. Do you remember that one ?

bthirion commented:
> also that would require also including the fsaverage white matter with the code -- but we may want to do that anyway.

Yes that's my opinion. It's pretty small.

jeromedockes commented:
related: #2313

jeromedockes commented:
> Also, there is a glitch that remains to be fixed. I could not find the issue, but it was something like a half voxel shift that needs to be addressed for freesurfer meshes. IIRC this resulted in LR inaccuracies typically. Do you remember that one ?

the neurostars question about reading surface files?

jeromedockes commented:
> Also, there is a glitch that remains to be fixed. I could not find the issue, but it was something like a half voxel shift that needs to be addressed for freesurfer meshes. IIRC this resulted in LR inaccuracies typically. Do you remember that one ?

do you mean that one: #2215

bthirion commented:
> also that would require also including the fsaverage white matter with the code -- but we may want to do that anyway.

Yes that's my opinion. It's pretty small.

bthirion commented:
> do you mean that one: #2215

Yes, exactly

jeromedockes commented:
bthirion <notifications@github.com> writes:

>> do you mean that one: #2215
>
> Yes, exactly

Indeed it hasn't been addressed. IIUC it's an issue with nibabel or
load_surf_mesh though, not related to vol_to_surf

jeromedockes commented:
> The problem is how to have both in the same API, and making it clear !

yes I agree. I seems that the list of depths would compete with the
`n_points` and `radius` parameters ... not sure what is the best way to
handle that, the new parameter would cause these to be ignored?

bthirion commented:
If a list of depthsis provided, `n_points` and `radius` would be ignored.

emdupre commented:
> where I can download white matter surface meshes for the fsaverage5 and fsaverage surfaces
to add fetchers for these.

These are included in the [main Freesurfer archive](https://surfer.nmr.mgh.harvard.edu/pub/dist/freesurfer/6.0.1/freesurfer-Linux-centos6_x86_64-stable-pub-v6.0.1.tar.gz) in the standard FS format (e.g. `rh.white`). They're also hosted on [OSF for the templateflow project](https://osf.io/ue5gx/) in the `fsaverage` folder at the 164k (fsaverage) and 10k (fsaverage5) resolutions.

@oesteban can confirm, but I'd imagine we could download directly from OSF !

jeromedockes commented:
> These are included in the [main Freesurfer archive](https://surfer.nmr.mgh.harvard.edu/pub/dist/freesurfer/6.0.1/freesurfer-Linux-centos6_x86_64-stable-pub-v6.0.1.tar.gz) in the standard FS format (e.g. `rh.white`). They're also hosted on [OSF for the templateflow project](https://osf.io/ue5gx/) in the `fsaverage` folder at the 164k (fsaverage) and 10k (fsaverage5) resolutions.
>
thanks! I saw this a bit late :/ but I did get them from the freesurfer
archive as you had suggested IRL. I converted them to gifti using
freesurfer's mris_convert tool
re. hosting: other fsaverage5 surfaces are shipped with nilearn so I
guess these should be as well, as suggested by @bthirion's comment.
for fsaverage, rather than getting it from a separate place, I was
thinking of adding these two surfaces to the archive containing the rest
currently used by the fetcher, which is on nitrc. we can later also add
a copy on osf in case download from nitrc fails.
however I was planning to wait for 
https://github.com/nilearn/nilearn/pull/2514
to be merged before updating the fetcher

jeromedockes commented:
I added the white matter surfaces in https://www.nitrc.org/frs/download.php/11807/fsaverage.tar.gz

bthirion commented:
Thx !

oesteban commented:
> @oesteban can confirm, but I'd imagine we could download directly from OSF !

Yes, or even better from S3. That would also buy you a couple of nice things:

- Maintenance through templateflow
- GIFTI format.

jeromedockes commented:
> Yes, or even better from S3. That would also buy you a couple of nice things:
>
> - Maintenance through templateflow
> - GIFTI format.

Thanks. for the moment it seemed easier to convert to gifti and upload
to the collection on nitrc that already contains the other fsaverage
meshes downloaded by the nilearn fetcher, but this is good to keep in
mind and we could add osf or S3 as a fallback solution when nitrc server
is down (or vice-versa)

oesteban commented:
I would love to see a better integration of templateflow with nilearn. If there's interest in that, I could work out a PR.

I'm aware though, that templateflow might be a bit too experimental for nilearn. Happy to hear other thoughts.

bthirion commented:
> I'm aware though, that templateflow might be a bit too experimental for nilearn. Happy to hear other thoughts.

You read in our thoughts --- at least mine.  It looks like a rather heavy dependency...

jeromedockes commented:
@bthirion I added fetchers for fsaverage5 and fsaverage white matter and removed the utility that generated low-resolution testing meshes and use fsaverage instead. I think your other comments had been addressed so if CI passes this will be ready for another review

bthirion reviewed (approved):
I don't find any issue. Thx !

bthirion commented:
Any other review ? @juhuntenburg ?

juhuntenburg commented:
Yep I can look at it within the next days if that helps

juhuntenburg commented:
I am running into an issue when running the new tests, which is probably not related to this PR, but maybe someone can help?
```
E       fixture 'tmp_path_factory' not found
>       available fixtures: cache, capfd, capsys, doctest_namespace, monkeypatch, pytestconfig, record_xml_property, recwarn, request_mocker, temp_nilearn_data_dir, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```

jeromedockes commented:
I think you need to update pytest

Julia Huntenburg <notifications@github.com> writes:

> I am running into an issue when running the new tests, which is probably not related to this PR, but maybe someone can help?
> ```
> E       fixture 'tmp_path_factory' not found
>>       available fixtures: cache, capfd, capsys, doctest_namespace, monkeypatch, pytestconfig, record_xml_property, recwarn, request_mocker, temp_nilearn_data_dir, tmpdir, tmpdir_factory
>>       use 'pytest --fixtures [testpath]' for help on them.
> ```
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/nilearn/nilearn/pull/2517#issuecomment-698995323

juhuntenburg commented:
First thing I notice, the plotting of the nested surfaces doesn't seem to really work, or at least it isn't very intuitive / helpful. Maybe the meshes could be rotated, or the inner one plotted with some transparency so that it becomes clearer?
![Figure_1](https://user-images.githubusercontent.com/6514016/94443655-54417000-019d-11eb-96c7-b5e1809bd190.png)
![Figure_4](https://user-images.githubusercontent.com/6514016/94443651-53104300-019d-11eb-9d0e-a4ac81f75387.png)

juhuntenburg commented on nilearn/_utils/data_gen.py:
```diff
@@ -8,6 +8,8 @@
 import numpy as np
 import pandas as pd
 import scipy.signal
+from scipy.spatial import Delaunay
```
Are you using thing anywhere?

jeromedockes commented:
Julia Huntenburg <notifications@github.com> writes:

> First thing I notice, the plotting of the nested surfaces doesn't seem to really work, or at least it isn't very intuitive / helpful. Maybe the meshes could be rotated, or the inner one plotted with some transparency so that it becomes clearer?
> ![Figure_1](https://user-images.githubusercontent.com/6514016/94443655-54417000-019d-11eb-96c7-b5e1809bd190.png)
> ![Figure_4](https://user-images.githubusercontent.com/6514016/94443651-53104300-019d-11eb-9d0e-a4ac81f75387.png)

I changed the camera position and made the outer one transparent. it is
still not super easy to see because matplotlib gets the overlays wrong
but I think a bit better than before. to really improve the plot the
solution would be to replace matplotlib with plotly in this example,
which we can do using nilearn's `view_surf` and `view_markers` -- but
this would best be addressed in a different issue and PR

jeromedockes commented:
>  import numpy as np
>  import pandas as pd
>  import scipy.signal
> +from scipy.spatial import Delaunay
>
> Are you using thing anywhere?

thanks, this was left after I removed some code -- I removed the imports now

jeromedockes commented:
circleci failure is unrelated and fixed in another PR

juhuntenburg commented on nilearn/surface/surface.py:
```diff
@@ -424,6 +482,8 @@ def vol_to_surf(img, surf_mesh,
     kind : {'line', 'ball'}
         The strategy used to sample image intensities around each vertex.
+        If `inner_mesh` is provided this parameter is ignored and values are
+        sampled along a line
```
Why not make a third kind 'depth' or similar? I find it not obvious otherwise that there is a third way to sample, which is invoked when inner_mesh is provided

jeromedockes replied:
I added the 'depth' strategy, and made the default 'auto', which chooses 'depth' if 'inner_mesh' is provided and 'line' otherwise

juhuntenburg reviewed (commented):
Looks good to me, the new way of plotting the surfaces helps clarity! I made some small comments. Most importantly, is there a reason why you chose not to indicate this new sampling method as a third 'kind' value? For me that would make it more obvious that there are three ways of sampling. The way it is implemented now (sampling between surfaces is chosen if inner surface is given) is less explicit. 

Also, I think it would be really great to have a real example for this, i.e. with real brain data, sampling at different depths of the cortex. This is a nice feature and would be worth showcasing. But maybe that is for another PR :)

jeromedockes commented on examples/01_plotting/plot_surface_projection_strategies.py:
```diff
@@ -11,10 +10,14 @@
-along the normal to the mesh, or inside a ball around the vertex. Don't worry
-too much about choosing one or the other: they take a similar amount of time
-and give almost identical results for most images.
+Three strategies are available to choose sample locations: they can be spread
```
thanks, I updated it with your suggestion

jeromedockes commented on examples/01_plotting/plot_surface_projection_strategies.py:
```diff
@@ -58,12 +65,33 @@
+# position of samples along the line
+
+nested_sample_points = surface._sample_locations_between_surfaces(
+    mesh, inner_mesh, np.eye(4), depth=[-.5, 0., .8, 1.])
```
yes, to show that it is possible to go outside the provided surfaces (ie negative values are allowed)

jeromedockes replied:
or values greater than 1 as well, I added 1.2

juhuntenburg reviewed (approved):
All concerns addressed for me. I would just want to check if others agree with referring to the new strategy as "depth" (@bthirion ?). I prefer this because I think it will be intuitive for many people. But I can also see an argument not to get into anatomical definitions. Although I think as long as we use depth and not anything related to layers it is ok

bthirion commented:
+1 for 'depth'
Merging.
Huge thanks to @jeromedockes
